File: android_app/app/src/main/java/com/oceansentinels/app/domain/repository/UserRepository.kt
package com.oceansentinels.app.domain.repository

import com.oceansentinels.app.domain.model.User
import com.oceansentinels.app.domain.model.UserRole

/**
 * Repository interface for user management operations (Admin only)
 */
interface UserRepository {
    
    /**
     * Get all users
     */
    suspend fun getAllUsers(): Result<List<User>>
    
    /**
     * Get users by role
     */
    suspend fun getUsersByRole(role: UserRole): Result<List<User>>
    
    /**
     * Get a single user by ID
     */
    suspend fun getUser(id: Int): Result<User>
    
    /**
     * Create a new user (Admin only)
     */
    suspend fun createUser(
        username: String,
        email: String,
        password: String,
        firstName: String,
        lastName: String,
        role: UserRole,
        phone: String? = null,
        location: String? = null
    ): Result<User>
    
    /**
     * Update user information
     */
    suspend fun updateUser(
        id: Int,
        firstName: String? = null,
        lastName: String? = null,
        phone: String? = null,
        location: String? = null,
        isActive: Boolean? = null
    ): Result<User>
    
    /**
     * Delete a user (Admin only)
     */
    suspend fun deleteUser(id: Int): Result<Unit>
    
    /**
     * Get rescue teams count
     */
    suspend fun getRescueTeamsCount(): Result<Int>
    
    /**
     * Get authorities count
     */
    suspend fun getAuthoritiesCount(): Result<Int>
}

File: IOS/OceanSentinels/Domain/Repositories/UserRepository.swift
import Foundation

/// Repository protocol for user management operations.
protocol UserRepository {

    /// Fetches all users (admin-only).
    func getAllUsers() async throws -> [User]

    /// Fetches users filtered by role.
    func getUsersByRole(role: UserRole) async throws -> [User]

    /// Fetches a single user by ID.
    func getUser(id: Int) async throws -> User

    /// Creates a new user with the specified role (admin).
    func createUser(
        username: String,
        email: String,
        password: String,
        firstName: String,
        lastName: String,
        role: UserRole,
        phone: String?,
        location: String?
    ) async throws -> User

    /// Creates a new user account as an admin (using RegistrationData).
    func adminCreateUser(data: RegistrationData) async throws -> User

    /// Updates a user's profile fields.
    func updateUser(
        id: Int,
        firstName: String?,
        lastName: String?,
        phone: String?,
        location: String?,
        isActive: Bool?
    ) async throws -> User

    /// Deletes a user by ID (admin-only).
    func deleteUser(id: Int) async throws

    /// Count of rescue team users.
    func getRescueTeamsCount() async throws -> Int

    /// Count of authority users.
    func getAuthoritiesCount() async throws -> Int
}
